### ISSUE TITLE
Add binding and tests for sparse_svm

### ISSUE THREAD

**Issue by rcurtin** [mlpack/mlpack#1633]
We have a nice implementation of a sparse SVM function from @shikharbhardwaj in `src/mlpack/methods/sparse_svm/`, but we actually don't have a nice binding to use it and it's not tested either.  This issue would be a nice task for someone who wants to learn more about the binding system and how to write new bindings.  To work on this, familiarity with SVMs will be necessary.  I'll try and write a basic description of what's involved here, but don't feel obligated to follow it to the letter---it is just a guide.

1.  Take a look at `src/mlpack/methods/sparse_svm/sparse_svm_function.hpp` (and `sparse_svm_function_impl.hpp`) to understand the function that is being optimized.  Basically this is just the linear SVM objective function.  Now take a look at the code in `src/mlpack/methods/softmax_regression/`---that directory has `softmax_regression_function.hpp` just like `sparse_svm_function.hpp`, but it also has a class `SoftmaxRegression` in `softmax_regression.hpp` which we lack for sparse SVMs.

2. Take a look at the way `SoftmaxRegression` is tested in `src/mlpack/tests/softmax_regression_test.cpp` (and also `LogisticRegression` and `DecisionTree` have relevant tests, since these are all classifiers).

3. The first step of coding here will actually be to create a `SparseSVM` class that is a lot like the `SoftmaxRegression` class (i.e. has basically the same method signatures), except that it will use `SparseSVMFunction` internally instead of `SoftmaxRegressionFunction`.  Note that you'll need to use ensmallen optimizers for this; specifically, you'll want to use a sparse differentiable separable optimizer type (http://ensmallen.org/docs.html#sparse-differentiable-separable-functions) by default, but the user should be able to choose the optimizer type they want.  It may also be useful to provide a function `SparseSVMFunction::EvaluateWithGradient()`---that can help make things faster.

4. Once we have a `SparseSVM` class, we'll need to write tests for it in the same way that the other classifiers have tests written.  Create a file `sparse_svm_test.cpp` with these tests in it, using the other tests that you looked at earlier as a guide.

5. Once `SparseSVM` works and is tested, take a look at the binding definitions in `src/mlpack/methods/logistic_regression/logistic_regression_main.cpp`, `src/mlpack/methods/softmax_regression/softmax_regression_main.cpp`, `src/mlpack/methods/decision_tree/decision_tree_main.cpp`, and `src/mlpack/methods/random_forest/random_forest_main.cpp`.  These are all other bindings for classifiers.  Create a new binding `src/mlpack/methods/sparse_svm/sparse_svm_main.cpp` that has the same general structure and can use the `SparseSVM` functionality to train a classifier, use the classifier to produce predictions (and probabilities) for new points, and save/load models.

6.  Update `src/mlpack/methods/sparse_svm/CMakeLists.txt` to build the binding; add `add_cli_executable(sparse_svm)` and `add_python_binding(sparse_svm)`, and then you will be able to `make mlpack_sparse_svm` from the command line for running it by hand.

7.  Look at the tests in `src/mlpack/tests/main_tests/` for inspiration and create a test for each of the options of `sparse_svm_main.cpp`.  Our goal in this test is not to test that the SVM gives correct output (that should be in the test you wrote in step 4), but instead to make sure that the input processing is reasonable and correct.  So, e.g., you should test that the program throws errors when invalid parameters are specified, and that it gives different results for different parameter sets; i.e., when using optimizers A and B, the predictions for a test set should be different (because the optimizers likely found different optima).

8.  Check the program for speed.  Compare with other SVM implementations (like libsvm or Shogun or scikit's implementation) and make sure that we can learn comparably fast (hopefully a lot faster!).  We'll need to make sure the implementation is fast before we release it.

9.  Bonus: (maybe this is good for a separate PR/issue or future work) extend `SparseSVMFunction` to handle kernelized SVMs, and update the binding and tests accordingly.  Actually this should not be that hard to do with a basic understand of kernelization and kernel SVMs, so maybe it would not be too much code to include in a PR.

**Comment by KimSangYeon-DGU** [comment]
@rcurtin I would like to work on this.

**Comment by rcurtin** [comment]
Sure, feel free.  This one will be a little more involved than #1625. :)

**Comment by AYESDIE** [comment]
hey @KimSangYeon-DGU, are you working on this issue? 
I would also love to work on this. 
I have a class SparseSVM ready, would you like to team up and split the task?

**Comment by KimSangYeon-DGU** [comment]
@AYESDIE Actually, I'm working on this.

**Comment by AYESDIE** [comment]
alright, have fun :)

**Comment by KimSangYeon-DGU** [comment]
@AYESDIE Sorry, for the my delayed PR, actually, I rethink how can I solve this. 
If you want to team up, how can we team up?? :)

**Comment by AYESDIE** [comment]
If you are on `discord`, I can email you my tag and we can talk, discuss and solve it there.

Else, we can discuss through email and solve it out.

**Comment by KimSangYeon-DGU** [comment]
@AYESDIE Is there any further progress??, If not, I want to solve this. Actually, I made a lot of tries for this.
I'm afraid I'm being too greedy.

**Comment by KimSangYeon-DGU** [comment]
@AYESDIE My email address is sy0814k@gmail.com

**Comment by KimSangYeon-DGU** [comment]
@AYESDIE I sent a request for communication in `discord`.

**Comment by KimSangYeon-DGU** [comment]
@AYESDIE Could you check your email or  `discord` if you have a time??

**Comment by KimSangYeon-DGU** [comment]
Because he did it first and I respect the spirit of the open source, I handed over this to @AYESDIE . Perhaps he will start working if he reads the mail. :)

**Comment by rcurtin** [comment]
The implementation and implementation tests are done now, but the binding and speed comparison is not done yet, so that's all that remain.

**Comment by Yashwants19** [comment]
Hi @rcurtin can I go binding here.
Thank you :)

**Comment by rcurtin** [comment]
@Yashwants19 if nobody else speaks up and says they are working on it, then sure, but maybe it's better to finish off the other PRs you have open first (I'm thinking of the GMM tests and the random forest tree depth changes, not the Go bindings---I haven't found any time to review those since it will probably take me 4 hours or so to go through it :)).

**Comment by Yashwants19** [comment]
Sure I will complete the present PRs and then I will start working on bindings. 
Thanks :)
